speechbrain/utils/distributed.py

```diff
@@ -51,25 +55,134 @@ def run_on_main(
         try:
             func(*args, **kwargs)
         finally:
-            sb.ddp_barrier()
+            ddp_barrier()
     else:
         # Others go here
-        sb.ddp_barrier()
+        ddp_barrier()
     if post_func is not None:
         if run_post_on_main:
             # Just run on every process without any barrier.
             post_func(*post_args, **post_kwargs)
-        elif not sb.if_main_process():
+        elif not if_main_process():
             # Others go here
             try:
                 post_func(*post_args, **post_kwargs)
             finally:
-                sb.ddp_barrier()
+                ddp_barrier()
         else:
             # But main comes here
-            sb.ddp_barrier()
+            ddp_barrier()
+
+
+def if_main_process():
+    """Check if the current process is the main process and authorized to run
+    I/O commands. In DDP mode, the main process is the one with RANK == 0.
+    In standard mode, the process will not have `RANK` Unix var and will be
+    authorized to run the I/O commands.
+    """
+    if "RANK" in os.environ:
+        if os.environ["RANK"] == "":
+            return False
+        else:
+            if int(os.environ["RANK"]) == 0:
+                return True
+            return False
+    return True
+
+
+def ddp_barrier():
+    """In DDP mode, this function will synchronizes all processes.
+    torch.distributed.barrier() will block processes until the whole
+    group enters this function.
+    """
+    if torch.distributed.is_initialized():
+        torch.distributed.barrier()
+
+
+def ddp_init_group(run_opts):
+    """
+    This function will initialize the ddp group if
+    distributed_launch=True bool is given in the python command line.
+
+    The ddp group will use distributed_backend arg for setting the
+    DDP communication protocol. `RANK` Unix variable will be used for
+    registring the subprocess to the ddp group.
+
+    Arguments
+    ---------
+    run_opts: list
+        a list of arguments to parse, most often from `sys.argv[1:]`
+    """
+    if run_opts["distributed_launch"]:
+        if "local_rank" not in run_opts:
+            raise ValueError(
+                "To use DDP backend, start your script with:\n\t"
+                "python -m torch.distributed.lunch [args]\n\t"
+                "experiment.py hyperparams.yaml --distributed_launch=True "
+                "--distributed_backend=nccl"
+            )
+        else:
+            if run_opts["local_rank"] + 1 > torch.cuda.device_count():
+                raise ValueError(
+                    "Killing process " + str() + "\n"
+                    "To use DDP backend, start your script with:\n\t"
+                    "python -m torch.distributed.lunch [args]\n\t"
+                    "experiment.py hyperparams.yaml --distributed_launch=True "
+                    "--distributed_backend=nccl"
+                )
+        if "RANK" in os.environ is None or os.environ["RANK"] == "":
```

Comment:
The return value of `"RANK" in os.environ`  is a boolean, which can never be `None`.

So
```python
"RANK" in os.environ is None
```
does not make sense.

Reply:
Where did you find that it is a boolean ? From what I see from the doc, it is a string.

Reply:
It is from https://docs.python.org/3/reference/datamodel.html#object.__contains__

---

You can write a test
```python
>>> isinstance('' in {}, bool)
True
```

`os.environ` is a dict. 

`"RANK" in os.environ` will invoke the `__contains__` method of the dict, which returns `True` or `False`.
It never returns a string, I believe.

---

> From what I see from the doc, it is a string.

Could you point me to the doc you are reading?

Reply:
https://docs.python.org/3/library/os.html

@aheba

Reply:
https://docs.python.org/3/library/os.html#os.environ
reads
> A mapping object representing the string environment.

which shows `os.environ` is indeed a dict, i.e., a mapping object. Cannot find where it says `os.environ` is a string or the result of
`"RANK" in os.environ`  is a string.

Reply:
Right, could you do a PR to fix this ? So we can test it :D